lib/puppet/util/pidlock.rb

```diff
@@ -67,7 +67,15 @@ def clear_if_stale
     elsif Puppet.features.microsoft_windows?
       # On Windows, we're checking if the filesystem path name of the running
       # process is our vendored ruby:
-      exe_path = Puppet::Util::Windows::Process::get_process_image_name_by_pid(lock_pid)
+      begin
+        exe_path = Puppet::Util::Windows::Process::get_process_image_name_by_pid(lock_pid)
+      rescue Puppet::Util::Windows::Error => e
+        # If we still don't have access to open the process token, we
+        # assume it's no longer a puppet process and clear the
+        # lockfile, otherwise we raise the error.
+        raise unless e.code == Puppet::Util::Windows::Error::ERROR_ACCESS_DENIED
+      end
+
       @lockfile.unlock unless exe_path =~ /\\bin\\ruby.exe$/
```

Comment:
The `Windows::Error` could be raised under a few different cases:

* If we're not running elevated (`Puppet.features.root?` is false), then `with_privilege` just yields. So the call to `open_process` can fail if the PID was started by a different user.
* If we are running elevated, but the `SeDebug` privilege isn't in our process token, then trying to enable the privilege can fail (we can only enable privileges that are disabled in our process token, but we can't add privileges we weren't granted).
* Even if we are running elevated, we may not have permission to open the process token. Try setting the pidfile to a privileged process like `csrss.exe`

I'm thinking if any of those cases raise, then we shouldn't unlock the lockfile, so something like:

```ruby
   begin
        exe_path = Puppet::Util::Windows::Process::get_process_image_name_by_pid(lock_pid)
        @lockfile.unlock unless exe_path =~ /\\bin\\ruby.exe$/
      rescue Puppet::Util::Windows::Error => e
        Puppet.debug("Failed to read pidfile #{file_path}: #{e.message}")                                                                  # If 
      end
```

Reply:
Makes sense. I tested with `csrss.exe` and thought it would be valid to unlock the lockfile in that case, but like you said there's a lot of stuff to take into account (not running elevated or no privilege), so maybe it's better not to unlock and let the user handle it. :+1: